File: incre-tests/synduce/list_to_tree/mls.f
Inductive Tree = empty Unit | node {Int, Tree, Tree};
Inductive List = nil Unit | cons {Int, List};

max = \a: Int. \b: Int. if (< a b) then b else a;

cat = fix (\f: List->List->List. \x: List. \y: List.
  match x with
    cons {h, t} -> cons {h, f t y}
  | nil _ -> y
  end
);

repr =
  let dec = fix (
    \f: List -> Tree -> List. \res: List. \t: Tree.
    match t with
      empty _ -> res
    | node {w, l, r} -> cons {w, f (f res l) r}
    end
  ) in fix (
    \f: Tree -> List. \t: Tree.
    match t with
      empty _ -> nil unit
    | node {w, l, r} -> cons {w, dec (f l) r}
    end
  );

target = fix (
  \f: Tree -> Compress Tree. \t: Tree.
  match t with
    empty _ -> empty unit
  | node {w, l, r} -> node {w, f l, f r}
  end
);

@Input x: Int;

spec = fix (
  \f: List -> {Int, Int}. \xs: List.
  match xs with
    nil _ -> {0,0}
  | cons {h, t} -> 
    let result = f t in
    {+ h result.1, max result.2 (+ h result.1)}
  end
);

main = \xs: Tree. spec (repr (target xs));

File: ml-benchmark/synduce/list_to_tree/mls.ml
type tree = Empty | Node of int * tree * tree
type list = Nil | Cons of int * list

let max a b = if a < b then b else a

let rec cat x y =
  match x with
  | Cons (h, t) -> Cons (h, cat t y)
  | Nil -> y

let rec dec res t =
  match t with
  | Empty -> res
  | Node (w, l, r) -> Cons (w, dec (dec res l) r)

let rec repr t =
  match t with
  | Empty -> Nil
  | Node (w, l, r) -> Cons (w, dec (repr l) r)

let rec target t =
  match t with
  | Empty -> Empty
  | Node (w, l, r) -> Node (w, target l, target r)

let x = 0

let rec spec xs =
  match xs with
  | Nil -> (0, 0)
  | Cons (h, t) ->
      match spec t with
      | result1, result2 ->
          let sum = h + result1 in
          (sum, max result2 sum)

let program xs = spec (repr (target xs))